src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockInventoryScreen.java

```diff
@@ -0,0 +1,129 @@
+import net.minecraft.nbt.NbtOps;
+import net.minecraft.nbt.StringNbtReader;
+import net.minecraft.nbt.visitor.StringNbtWriter;
+import net.minecraft.screen.slot.Slot;
+import net.minecraft.screen.slot.SlotActionType;
+import net.minecraft.util.Identifier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Opened here {@code de.hysky.skyblocker.mixins.MinecraftClientMixin#skyblocker$skyblockInventoryScreen}
+ * <br>
+ * Book button is moved here {@code de.hysky.skyblocker.mixins.InventoryScreenMixin#skyblocker}
+ */
+public class SkyblockInventoryScreen extends InventoryScreen {
+    private static final Logger LOGGER = LoggerFactory.getLogger("Equipment");
+
+    public static final ItemStack[] equipment = new ItemStack[]{ItemStack.EMPTY, ItemStack.EMPTY, ItemStack.EMPTY, ItemStack.EMPTY};
+    private static final Codec<ItemStack[]> CODEC = ItemUtils.EMPTY_ALLOWING_ITEMSTACK_CODEC.listOf(4,4)
+            .xmap(itemStacks -> itemStacks.toArray(ItemStack[]::new), List::of).fieldOf("items").codec();
+
+    private static final Identifier SLOT_TEXTURE = new Identifier("container/slot");
+    private static final Identifier EMPTY_SLOT = new Identifier(SkyblockerMod.NAMESPACE, "equipment/empty_icon");
+
+    public static void initEquipment() {
+        ClientLifecycleEvents.CLIENT_STARTED.register(client1 -> {
+            Path resolve = SkyblockerMod.CONFIG_DIR.resolve("equipment.nbt");
+            try (BufferedReader reader = Files.newBufferedReader(resolve)) {
+                ItemStack[] array = CODEC.parse(
+                                NbtOps.INSTANCE, StringNbtReader.parse(reader.lines().collect(Collectors.joining())))
+                        .getOrThrow();
+                System.arraycopy(array, 0, equipment, 0, Math.min(array.length, 4));
+            } catch (Exception e) {
+                LOGGER.error("[Skyblocker] Failed to load Equipment data", e);
+            }
+        });
+
+        ClientLifecycleEvents.CLIENT_STOPPING.register(client1 -> {
+            Path resolve = SkyblockerMod.CONFIG_DIR.resolve("equipment.nbt");
+            try (BufferedWriter writer = Files.newBufferedWriter(resolve)) {
+                writer.write(new StringNbtWriter().apply(CODEC.encodeStart(NbtOps.INSTANCE, equipment).getOrThrow()));
+            } catch (Exception e) {
+                LOGGER.error("[Skyblocker] Failed to save Equipment data", e);
+            }
+        });
+    }
+
+    public SkyblockInventoryScreen(PlayerEntity player) {
+        super(player);
+        SimpleInventory inventory = new SimpleInventory(equipment);
+
+        Slot slot = handler.slots.get(45);
+        ((SlotAccessor) slot).setX(slot.x + 21);
+        for (int i = 0; i < 4; i++) {
+            handler.slots.add(new EquipmentSlot(inventory, i, 77, 8 + i * 18));
```

Comment:
I really don't like this. Since the only thing this is doing is rendering it for you, you could just not add it to the handler and render it on the screen yourself.

Reply:
ik it looks kinda wack but it makes my life a whole lot easier, it does the item's rendering, tooltip and the white square thing you hover it for me! Also there's no way for it to send anything to the server i triple checked. But if you really don't like it then I can change it

Reply:
I'm more worried that mojang might change this behavior in the future.

Reply:
Then we shall catch it during snapshots. Also even if i do that I will need to keep these lines to move the offhand slot anyways:
```java
Slot slot = handler.slots.get(45);
((SlotAccessor) slot).setX(slot.x + 21);
```

Reply:
I'm still not convinced. I checked and `slots` is used as the "official" list of slots for the client side and is used in many instances to handle events before sending to the server, such as `HandledScreen#mouseClicked`, the interaction manager, etc. The mouse click code in `HandledScreen` is a bit scary as it treats slots as movable stacks and it gets called before your custom `SkyblockInventoryScreen#onMouseClick`.

You can render the slot with `HandledScreen#drawSlot`, highlight with `HandledScreen#drawSlotHighlight`, and tooltip similar to `AuctionViewScreen#drawMouseoverTooltip`.

Imo modifying slots is fine, which is what you're doing with the offhand, as with fancy auction screens, etc, but adding slots is dangerous. Please change while also fixing biglo's review, or I can if you don't want to.